src/x86/intel_joule_expansion.c

```diff
@@ -24,48 +25,83 @@
 
 #include <stdlib.h>
 #include <string.h>
+#include <fcntl.h>
 #include <sys/types.h>
 #include <sys/stat.h>
-#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <linux/i2c-dev.h>
+
+// Debug
+#include <syslog.h>
+#include <sys/errno.h>
 
 #include "common.h"
 #include "x86/intel_joule_expansion.h"
 
 #define PLATFORM_NAME "INTEL JOULE EXPANSION"
 
+typedef union i2c_smbus_data_union {
+    uint8_t byte;        ///< data byte
+    unsigned short word; ///< data short word
+    uint8_t block[I2C_SMBUS_BLOCK_MAX + 2];
+    ///< block[0] is used for length and one more for PEC
+} i2c_smbus_data_t;
+
+typedef struct i2c_smbus_ioctl_data_struct {
+    uint8_t read_write;     ///< operation direction
+    uint8_t command;        ///< ioctl command
+    int size;               ///< data size
+    i2c_smbus_data_t* data; ///< data
+} i2c_smbus_ioctl_data_t;
+
+typedef enum {NO_SHIELD, DFROBOT, GROVE} shield_t;
+
+static mraa_result_t
+mraa_joule_expansion_board_get_valid_fp(mraa_aio_context dev)
+{
+    char file_path[64] = "";
+
+    // 4 channels per ADC
+    snprintf(file_path, 64, "/sys/bus/iio/devices/iio:device%d/in_voltage%d_raw", dev->channel / 4, dev->channel % 4);
```

Comment:
The aio module doesn't handle multiple devices. In fact, it assumes iio:device0 is the ADC. Proper way would be to actually query all the IIO devices and figure out which ones are the ADCs. That's a pretty substantial "fix".